File: onedal/svm/svm.py
# ==============================================================================
# Copyright 2021 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from abc import ABCMeta, abstractmethod

import numpy as np
from scipy import sparse as sp

from onedal import _backend

from ..common._base import BaseEstimator
from ..common._estimator_checks import _check_is_fitted
from ..common._mixin import ClassifierMixin, RegressorMixin
from ..common._policy import _get_policy
from ..datatypes import _convert_to_supported, from_table, to_table
from ..utils import (
    _check_array,
    _check_n_features,
    _check_X_y,
    _column_or_1d,
    _validate_targets,
)


class BaseSVM(BaseEstimator, metaclass=ABCMeta):
    @abstractmethod
    def __init__(
        self,
        C,
        nu,
        epsilon,
        kernel="rbf",
        *,
        degree,
        gamma,
        coef0,
        tol,
        shrinking,
        cache_size,
        max_iter,
        tau,
        class_weight,
        decision_function_shape,
        break_ties,
        algorithm,
    ):
        self.C = C
        self.nu = nu
        self.epsilon = epsilon
        self.kernel = kernel
        self.degree = degree
        self.coef0 = coef0
        self.gamma = gamma
        self.tol = tol
        self.shrinking = shrinking
        self.cache_size = cache_size
        self.max_iter = max_iter
        self.tau = tau
        self.class_weight = class_weight
        self.decision_function_shape = decision_function_shape
        self.break_ties = break_ties
        self.algorithm = algorithm

    def _validate_targets(self, y, dtype):
        self.class_weight_ = None
        self.classes_ = None
        return _column_or_1d(y, warn=True).astype(dtype, copy=False)

    def _get_onedal_params(self, dtype):
        max_iter = 10000 if self.max_iter == -1 else self.max_iter
        # TODO: remove this workaround
        # when oneDAL SVM starts support of 'n_iterations' result
        self.n_iter_ = 1 if max_iter < 1 else max_iter
        class_count = 0 if self.classes_ is None else len(self.classes_)
        return {
            "fptype": dtype,
            "method": self.algorithm,
            "kernel": self.kernel,
            "c": self.C,
            "nu": self.nu,
            "epsilon": self.epsilon,
            "class_count": class_count,
            "accuracy_threshold": self.tol,
            "max_iteration_count": int(max_iter),
            "scale": self._scale_,
            "sigma": self._sigma_,
            "shift": self.coef0,
            "degree": self.degree,
            "tau": self.tau,
            "shrinking": self.shrinking,
            "cache_size": self.cache_size,
        }

    def _fit(self, X, y, sample_weight, module, queue):
        if hasattr(self, "decision_function_shape"):
            if self.decision_function_shape not in ("ovr", "ovo", None):
                raise ValueError(
                    f"decision_function_shape must be either 'ovr' or 'ovo', "
                    f"got {self.decision_function_shape}."
                )

        X, y = _check_X_y(
            X,
            y,
            dtype=[np.float64, np.float32],
            force_all_finite=True,
            accept_sparse="csr",
        )
        # hard work remains on moving validate targets away from onedal
        y = self._validate_targets(y, X.dtype)
        if sample_weight is not None and len(sample_weight) > 0:
            sample_weight = _check_array(
                sample_weight,
                accept_sparse=False,
                ensure_2d=False,
                dtype=X.dtype,
                order="C",
            )
        elif self.class_weight is not None:
            sample_weight = np.ones(X.shape[0], dtype=X.dtype)

        if sample_weight is not None:
            if self.class_weight_ is not None:
                for i, v in enumerate(self.class_weight_):
                    sample_weight[y == i] *= v
            data = (X, y, sample_weight)
        else:
            data = (X, y)
        self._sparse = sp.issparse(X)

        if self.kernel == "linear":
            self._scale_, self._sigma_ = 1.0, 1.0
            self.coef0 = 0.0
        else:
            self._scale_, self._sigma_ = self.gamma, np.sqrt(0.5 / self.gamma)

        policy = _get_policy(queue, *data)
        data_t = to_table(*_convert_to_supported(policy, *data))
        params = self._get_onedal_params(data_t[0].dtype)
        result = module.train(policy, params, *data_t)

        if self._sparse:
            self.dual_coef_ = sp.csr_matrix(from_table(result.coeffs).T)
            self.support_vectors_ = sp.csr_matrix(from_table(result.support_vectors))
        else:
            self.dual_coef_ = from_table(result.coeffs).T
            self.support_vectors_ = from_table(result.support_vectors)

        self.intercept_ = from_table(result.biases).ravel()
        self.support_ = from_table(result.support_indices).ravel().astype("int")
        self.n_features_in_ = X.shape[1]
        self.shape_fit_ = X.shape

        # _n_support not used in this object, will be moved to sklearnex
        if getattr(self, "classes_", None) is not None:
            indices = y.take(self.support_, axis=0)
            self._n_support = np.array(
                [np.sum(indices == i) for i, _ in enumerate(self.classes_)]
            )
        self._gamma = self._scale_

        self._onedal_model = result.model
        return self

    def _create_model(self, module):
        m = module.model()

        m.support_vectors = to_table(self.support_vectors_)
        m.coeffs = to_table(self.dual_coef_.T)
        m.biases = to_table(self.intercept_)
        return m

    def _infer(self, X, module, queue):
        _check_is_fitted(self)

        if self._sparse:
            if not sp.isspmatrix(X):
                X = sp.csr_matrix(X)
            else:
                X.sort_indices()
        elif sp.issparse(X) and not callable(self.kernel):
            raise ValueError(
                "cannot use sparse input in %r trained on dense data"
                % type(self).__name__
            )

        policy = _get_policy(queue, X)
        X = to_table(_convert_to_supported(policy, X))
        params = self._get_onedal_params(X.dtype)

        if hasattr(self, "_onedal_model"):
            model = self._onedal_model
        else:
            model = self._create_model(module)
        return module.infer(policy, params, model, X)

    def _predict(self, X, module, queue):
        return from_table(self._infer(X, module, queue).responses)

    def _decision_function(self, X, module, queue):
        return from_table(self._infer(X, module, queue).decision_function)


class SVR(RegressorMixin, BaseSVM):
    """
    Epsilon--Support Vector Regression.
    """

    def __init__(
        self,
        C=1.0,
        epsilon=0.1,
        kernel="rbf",
        *,
        degree=3,
        gamma="scale",
        coef0=0.0,
        tol=1e-3,
        shrinking=True,
        cache_size=200.0,
        max_iter=-1,
        tau=1e-12,
        algorithm="thunder",
    ):
        super().__init__(
            C=C,
            nu=0.5,
            epsilon=epsilon,
            kernel=kernel,
            degree=degree,
            gamma=gamma,
            coef0=coef0,
            tol=tol,
            shrinking=shrinking,
            cache_size=cache_size,
            max_iter=max_iter,
            tau=tau,
            class_weight=None,
            decision_function_shape=None,
            break_ties=False,
            algorithm=algorithm,
        )

    def fit(self, X, y, sample_weight=None, queue=None):
        return super()._fit(X, y, sample_weight, _backend.svm.regression, queue)

    def predict(self, X, queue=None):
        return super()._predict(X, _backend.svm.regression, queue)


class SVC(ClassifierMixin, BaseSVM):
    """
    C-Support Vector Classification.
    """

    def __init__(
        self,
        C=1.0,
        kernel="rbf",
        *,
        degree=3,
        gamma="scale",
        coef0=0.0,
        tol=1e-3,
        shrinking=True,
        cache_size=200.0,
        max_iter=-1,
        tau=1e-12,
        class_weight=None,
        decision_function_shape="ovr",
        break_ties=False,
        algorithm="thunder",
    ):
        super().__init__(
            C=C,
            nu=0.5,
            epsilon=0.0,
            kernel=kernel,
            degree=degree,
            gamma=gamma,
            coef0=coef0,
            tol=tol,
            shrinking=shrinking,
            cache_size=cache_size,
            max_iter=max_iter,
            tau=tau,
            class_weight=class_weight,
            decision_function_shape=decision_function_shape,
            break_ties=break_ties,
            algorithm=algorithm,
        )

    def _create_model(self, module):
        m = super()._create_model(module)
        m.first_class_response, m.second_class_response = 0, 1
        return m

    def _validate_targets(self, y, dtype):
        y, self.class_weight_, self.classes_ = _validate_targets(
            y, self.class_weight, dtype
        )
        return y

    def fit(self, X, y, sample_weight=None, queue=None):
        return super()._fit(X, y, sample_weight, _backend.svm.classification, queue)

    def predict(self, X, queue=None):
        return super()._predict(X, _backend.svm.classification, queue)

    def decision_function(self, X, queue=None):
        return super()._decision_function(X, _backend.svm.classification, queue)


class NuSVR(RegressorMixin, BaseSVM):
    """
    Nu-Support Vector Regression.
    """

    def __init__(
        self,
        nu=0.5,
        C=1.0,
        kernel="rbf",
        *,
        degree=3,
        gamma="scale",
        coef0=0.0,
        tol=1e-3,
        shrinking=True,
        cache_size=200.0,
        max_iter=-1,
        tau=1e-12,
        algorithm="thunder",
    ):
        super().__init__(
            C=C,
            nu=nu,
            epsilon=0.0,
            kernel=kernel,
            degree=degree,
            gamma=gamma,
            coef0=coef0,
            tol=tol,
            shrinking=shrinking,
            cache_size=cache_size,
            max_iter=max_iter,
            tau=tau,
            class_weight=None,
            decision_function_shape=None,
            break_ties=False,
            algorithm=algorithm,
        )

    def fit(self, X, y, sample_weight=None, queue=None):
        return super()._fit(X, y, sample_weight, _backend.svm.nu_regression, queue)

    def predict(self, X, queue=None):
        return super()._predict(X, _backend.svm.nu_regression, queue)


class NuSVC(ClassifierMixin, BaseSVM):
    """
    Nu-Support Vector Classification.
    """

    def __init__(
        self,
        nu=0.5,
        kernel="rbf",
        *,
        degree=3,
        gamma="scale",
        coef0=0.0,
        tol=1e-3,
        shrinking=True,
        cache_size=200.0,
        max_iter=-1,
        tau=1e-12,
        class_weight=None,
        decision_function_shape="ovr",
        break_ties=False,
        algorithm="thunder",
    ):
        super().__init__(
            C=1.0,
            nu=nu,
            epsilon=0.0,
            kernel=kernel,
            degree=degree,
            gamma=gamma,
            coef0=coef0,
            tol=tol,
            shrinking=shrinking,
            cache_size=cache_size,
            max_iter=max_iter,
            tau=tau,
            class_weight=class_weight,
            decision_function_shape=decision_function_shape,
            break_ties=break_ties,
            algorithm=algorithm,
        )

    def _create_model(self, module):
        m = super()._create_model(module)
        m.first_class_response, m.second_class_response = 0, 1
        return m

    def _validate_targets(self, y, dtype):
        y, self.class_weight_, self.classes_ = _validate_targets(
            y, self.class_weight, dtype
        )
        return y

    def fit(self, X, y, sample_weight=None, queue=None):
        return super()._fit(X, y, sample_weight, _backend.svm.nu_classification, queue)

    def predict(self, X, queue=None):
        return super()._predict(X, _backend.svm.nu_classification, queue)

    def decision_function(self, X, queue=None):
        return super()._decision_function(X, _backend.svm.nu_classification, queue)

File: onedal/svm/svm.cpp
/*******************************************************************************
* Copyright 2021 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

#include "oneapi/dal/algo/svm.hpp"

#include "onedal/common.hpp"
#include "onedal/primitives/kernel_functions.hpp"

namespace py = pybind11;

namespace oneapi::dal::python {

namespace svm {

template <typename Task, typename Ops>
struct method2t {
    method2t(const Task& task, const Ops& ops) : ops(ops) {}

    template <typename Float>
    auto operator()(const py::dict& params) {
        using namespace dal::svm;

        const auto method = params["method"].cast<std::string>();
        if constexpr (std::is_same_v<Task, task::classification>) {
            ONEDAL_PARAM_DISPATCH_VALUE(method, "thunder", ops, Float, method::thunder);
            ONEDAL_PARAM_DISPATCH_VALUE(method, "smo", ops, Float, method::smo);
            ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
        }
        else {
            ONEDAL_PARAM_DISPATCH_VALUE(method, "thunder", ops, Float, method::thunder);
            ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
        }
    }

    Ops ops;
};

template <typename Ops>
struct kernel2t {
    kernel2t(const Ops& ops) : ops(ops) {}

    template <typename Float, typename Method>
    auto operator()(const py::dict& params) {
        using namespace dal::svm;

        auto kernel = params["kernel"].cast<std::string>();
        ONEDAL_PARAM_DISPATCH_VALUE(kernel,
                                    "linear",
                                    ops,
                                    Float,
                                    Method,
                                    linear_kernel::descriptor<Float>);
        ONEDAL_PARAM_DISPATCH_VALUE(kernel,
                                    "rbf",
                                    ops,
                                    Float,
                                    Method,
                                    rbf_kernel::descriptor<Float>);
        ONEDAL_PARAM_DISPATCH_VALUE(kernel,
                                    "poly",
                                    ops,
                                    Float,
                                    Method,
                                    polynomial_kernel::descriptor<Float>);
        ONEDAL_PARAM_DISPATCH_VALUE(kernel,
                                    "sigmoid",
                                    ops,
                                    Float,
                                    Method,
                                    sigmoid_kernel::descriptor<Float>);
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(kernel);
    }

    Ops ops;
};

struct params2desc {
    template <typename Float, typename Method, typename Task, typename Kernel>
    auto operator()(const pybind11::dict& params) {
        using namespace dal::svm;

        constexpr bool is_cls = std::is_same_v<Task, task::classification>;
        constexpr bool is_nu_cls = std::is_same_v<Task, task::nu_classification>;
        constexpr bool is_reg = std::is_same_v<Task, task::regression>;
        constexpr bool is_nu_reg = std::is_same_v<Task, task::nu_regression>;

        auto desc = descriptor<Float, Method, Task, Kernel>{ get_kernel_descriptor<Kernel>(params) }
                        .set_max_iteration_count(params["max_iteration_count"].cast<std::int64_t>())
                        .set_accuracy_threshold(params["accuracy_threshold"].cast<double>())
                        .set_cache_size(params["cache_size"].cast<double>())
                        .set_tau(params["tau"].cast<double>())
                        .set_shrinking(params["shrinking"].cast<bool>());

        if constexpr (is_cls || is_reg || is_nu_reg) {
            desc.set_c(params["c"].cast<double>());
        }
        if constexpr (is_cls || is_nu_cls) {
            desc.set_class_count(params["class_count"].cast<std::int64_t>());
        }
        if constexpr (is_reg) {
            desc.set_epsilon(params["epsilon"].cast<double>());
        }
        if constexpr (is_nu_reg || is_nu_cls) {
            desc.set_nu(params["nu"].cast<double>());
        }

        return desc;
    }
};

template <typename Policy, typename Task>
void init_train_ops(py::module_& m) {
    m.def("train",
          [](const Policy& policy,
             const py::dict& params,
             const table& data,
             const table& responses,
             const table& weights) {
              using namespace dal::svm;
              using input_t = train_input<Task>;

              train_ops ops(policy, input_t{ data, responses, weights }, params2desc{});
              return fptype2t{ method2t{ Task{}, kernel2t{ ops } } }(params);
          });
}

template <typename Policy, typename Task>
void init_infer_ops(py::module_& m) {
    m.def("infer",
          [](const Policy& policy,
             const py::dict& params,
             const dal::svm::model<Task>& model,
             const table& data) {
              using namespace dal::svm;
              using input_t = infer_input<Task>;

              infer_ops ops(policy, input_t{ model, data }, params2desc{});
              return fptype2t{ method2t{ Task{}, kernel2t{ ops } } }(params);
          });
}

template <typename Task>
void init_model(py::module_& m) {
    using namespace dal::svm;
    using model_t = model<Task>;

    auto cls =
        py::class_<model_t>(m, "model")
            .def(py::init())
            .def(py::pickle(
                [](const model_t& m) {
                    return serialize(m);
                },
                [](const py::bytes& bytes) {
                    return deserialize<model_t>(bytes);
                }))
            .def_property_readonly("support_vector_count", &model_t::get_support_vector_count)
            .DEF_ONEDAL_PY_PROPERTY(support_vectors, model_t)
            .DEF_ONEDAL_PY_PROPERTY(coeffs, model_t)
            .DEF_ONEDAL_PY_PROPERTY(biases, model_t);

    constexpr bool is_classification = std::is_same_v<Task, task::classification>;
    constexpr bool is_nu_classification = std::is_same_v<Task, task::nu_classification>;

    if constexpr (is_classification || is_nu_classification) {
        // workaround for gcc which cannot deduce setters directly passed to def_property()
        auto class_1_setter = &model_t::template set_first_class_response<>;
        auto class_2_setter = &model_t::template set_second_class_response<>;

        cls.def_property("first_class_response",
                         &model_t::get_first_class_response,
                         class_1_setter);
        cls.def_property("second_class_response",
                         &model_t::get_second_class_response,
                         class_2_setter);
    }
}

template <typename Task>
void init_train_result(py::module_& m) {
    using namespace dal::svm;
    using result_t = train_result<Task>;

    py::class_<result_t>(m, "train_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(model, result_t)
        .DEF_ONEDAL_PY_PROPERTY(support_vectors, result_t)
        .DEF_ONEDAL_PY_PROPERTY(support_indices, result_t)
        .DEF_ONEDAL_PY_PROPERTY(coeffs, result_t)
        .DEF_ONEDAL_PY_PROPERTY(biases, result_t);
}

template <typename Task>
void init_infer_result(py::module_& m) {
    using namespace dal::svm;
    using result_t = infer_result<Task>;

    auto cls = py::class_<result_t>(m, "infer_result")
                   .def(py::init())
                   .DEF_ONEDAL_PY_PROPERTY(responses, result_t);

    constexpr bool is_classification = std::is_same_v<Task, task::classification>;
    constexpr bool is_nu_classification = std::is_same_v<Task, task::nu_classification>;

    if constexpr (is_classification || is_nu_classification) {
        // workaround for gcc which cannot deduce setters directly passed to def_property()
        auto setter = &result_t::template set_decision_function<>;
        cls.def_property("decision_function", &result_t::get_decision_function, setter);
    }
}

ONEDAL_PY_DECLARE_INSTANTIATOR(init_model);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_ops);

} // namespace svm

ONEDAL_PY_INIT_MODULE(svm) {
    using namespace svm;
    using namespace dal::svm;
    using namespace dal::detail;

    using task_list =
        types<task::classification, task::regression, task::nu_classification, task::nu_regression>;
    auto sub = m.def_submodule("svm");

    #ifndef ONEDAL_DATA_PARALLEL_SPMD
        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
        ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);

        ONEDAL_PY_INSTANTIATE(init_model, sub, task_list);
        ONEDAL_PY_INSTANTIATE(init_train_result, sub, task_list);
        ONEDAL_PY_INSTANTIATE(init_infer_result, sub, task_list);
    #endif
}

ONEDAL_PY_TYPE2STR(dal::svm::task::classification, "classification");
ONEDAL_PY_TYPE2STR(dal::svm::task::regression, "regression");
ONEDAL_PY_TYPE2STR(dal::svm::task::nu_classification, "nu_classification");
ONEDAL_PY_TYPE2STR(dal::svm::task::nu_regression, "nu_regression");

} // namespace oneapi::dal::python
